robosat/osm/road.py

```diff
@@ -0,0 +1,179 @@
+            "right_hard_shoulder_width": 0.75,
+        },
+        "secondary_link": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.75,
+        },
+        "tertiary_link": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.00,
+        },
+    }
+
+    road_filter = set(highway_attributes.keys())
+
+    EARTH_MEAN_RADIUS = 6371004.0
+
+    def __init__(self):
+        super().__init__()
+        self.features = []
+
+    def way(self, w):
+        if "highway" not in w.tags:
+            return
+
+        if w.tags["highway"] not in self.road_filter:
+            return
+
+        left_hard_shoulder_width = self.highway_attributes[w.tags["highway"]][
+            "left_hard_shoulder_width"
+        ]
+        lane_width = self.highway_attributes[w.tags["highway"]]["lane_width"]
+        lanes = self.highway_attributes[w.tags["highway"]]["lanes"]
+        right_hard_shoulder_width = self.highway_attributes[w.tags["highway"]][
+            "right_hard_shoulder_width"
+        ]
+
+        if "oneway" not in w.tags:
+            lanes = lanes * 2
+        elif w.tags["oneway"] == "no":
+            lanes = lanes * 2
+
+        if "lanes" in w.tags:
+            try:
+                lanes = int(w.tags["lanes"])
+            except ValueError as e:
+                print(
+                    "ValueError: ",
+                    e,
+                    " Invalid road lanes property: https://www.openstreetmap.org/way/{}".format(
+                        w.id
+                    ),
+                    file=sys.stderr,
+                )
+
+        road_width = (
+            left_hard_shoulder_width + lane_width * lanes + right_hard_shoulder_width
+        )
+
+        if "width" in w.tags:
+            try:
+                road_width = float(w.tags["width"])
+            except ValueError as e:
+                print(
+                    "ValueError: ",
+                    e,
+                    " Invalid road width property: https://www.openstreetmap.org/way/{}".format(
+                        w.id
+                    ),
+                    file=sys.stderr,
+                )
+
+        geometry = geojson.LineString([(n.lon, n.lat) for n in w.nodes])
+        shape = shapely.geometry.shape(geometry)
+        geometry_buffer = shape.buffer(
+            math.degrees(road_width / 2.0 / self.EARTH_MEAN_RADIUS)
+        )
+
+        if shape.is_valid:
+            feature = geojson.Feature(
+                geometry=shapely.geometry.mapping(geometry_buffer)
+            )
+            self.features.append(feature)
+        else:
+            print(
+                "Warning: invalid feature: https://www.openstreetmap.org/way/{}".format(
+                    w.id
+                ),
+                file=sys.stderr,
+            )
```

Comment:
Hm did you run black with 

    black . --exclude .env --check --line-length 120

this looks a bit squashed.

Reply:
Maybe when I first used Black plugin at VS Code, I didn't configure the "black.lineLength" value(default 88). After I configured it, it has changed the files already. But I rectified it in this commit.